codeface_extraction/extractions.py

```diff
@@ -266,6 +266,47 @@ def __init__(self, dbm, conf, resdir, csv_writer):
                     # LIMIT 10
                 """
 
+# Extraction of function implementations
+class FunctionImplementationExtraction(Extraction):
+    def __init__(self, dbm, conf, resdir, csv_writer):
+        Extraction.__init__(self, dbm, conf, resdir, csv_writer)
+
+        self.file_name = "implementations.list"
+
+        # for subclasses
+        self.sql = """
+                    SELECT c.id,
+                           c.commitHash, c.ChangedFiles,
+                           cd.file, cd.entityId, cd.impl
+
+                    FROM project p
+
+                    # get commits for project
+                    JOIN commit c ON p.id = c.projectId
+
+                    # get commit meta-data
+                    LEFT JOIN commit_dependency cd ON c.id = cd.commitId
+
+                    # filter for current project
+                    WHERE p.name = '{project}'
+                    AND p.analysisMethod = '{tagging}'
+                    AND cd.file IS NOT NULL
+
+                    ORDER BY cd.file, c.id, c.commitHash, cd.entityId
+                """
+
+    def _reduce_result(self, result):
+        newResult = []
+        for (commitId, commitHash, changedFiles, fileId, entityId, impl) in result:
+            newImpl = impl.replace("\n", " ")
+            newImpl = newImpl.replace("\r", " ")
+            newImpl = newImpl.replace("\t", " ")
+            newImpl = newImpl.replace(u"\ufffd", " ")
+
```

Comment:
instead of lines 301 - 304 I sugguest the following (NOTE: requires `import unicodedata, re`):
"""  Removes control characters such as \r\n \x1b \ufffd from string impl and returns a unicode string where all control characters have been replaced by a space """

    newImpl = re.sub(r'\\ufff.', ' ', impl)
    newImpl = "".join(ch if unicodedata.category(ch)[0]!="C" else " " for ch in newImpl.decode('unicode-escape'))

Reply:
This looks great. Thanks, @nischikata! Didn't know that there is a good approach for doing it in a more generic way. I just would like to watch for consistent quotation marks and spaces around operators.

I propose to give also more explanatory comments, such as giving the following links:
- https://www.fileformat.info/info/unicode/category/index.htm and
- https://www.compart.com/en/unicode/block/U+FFF0.

Reply:
I tested this and it is unfortunately not working as expected: The output gets scrambled and control characters are not removed:

Input: `öffi �￼˜`
Output: `Ã¶ffi ï¿½ï¿¼Ë`
Expected output: <code>öffi&nbsp;&nbsp;&nbsp;&nbsp;</code>

I talked to @nischikata and we agree that there is something going wrong already when writing the code into the database. We also agree that we should be input-tolerant and support the following both scenarios: 
1. The database contents are correctly encoded and we should be able to remove problematic characters.
2. The database contents are scrambled already and we should still be able to remove problematic characters.

We should talk about this sometime soon.